Source/Core/VideoBackends/Software/Rasterizer.cpp

```diff
@@ -20,28 +20,92 @@ namespace Rasterizer
 {
 static constexpr int BLOCK_SIZE = 2;
 
+struct SlopeContext
+{
+  SlopeContext(const OutputVertexData* v0, const OutputVertexData* v1, const OutputVertexData* v2,
+               s32 x0, s32 y0)
+      : x0(x0), y0(y0)
+  {
+    // adjust a little less than 0.5
+    const float adjust = 0.495f;
+
+    xOff = ((float)x0 - v0->screenPosition.x) + adjust;
+    yOff = ((float)y0 - v0->screenPosition.y) + adjust;
+
+    dx10 = v1->screenPosition.x - v0->screenPosition.x;
+    dx20 = v2->screenPosition.x - v0->screenPosition.x;
+    dy10 = v1->screenPosition.y - v0->screenPosition.y;
+    dy20 = v2->screenPosition.y - v0->screenPosition.y;
+  }
+  s32 x0;
+  s32 y0;
+  float xOff;
+  float yOff;
+  float dx10;
+  float dx20;
+  float dy10;
+  float dy20;
+};
+
+struct Slope
+{
+  Slope() = default;
+  Slope(float f0, float f1, float f2, const SlopeContext& ctx) : f0(f0)
+  {
+    float delta_20 = f2 - f0;
+    float delta_10 = f1 - f0;
+
+    //        x2 - x0    y1 - y0    x1 - x0    y2 - y0
+    float a = delta_20 * ctx.dy10 - delta_10 * ctx.dy20;
+    float b = ctx.dx20 * delta_10 - ctx.dx10 * delta_20;
+    float c = ctx.dx20 * ctx.dy10 - ctx.dx10 * ctx.dy20;
+
+    dfdx = a / c;
+    dfdy = b / c;
+
+    x0 = ctx.x0;
+    y0 = ctx.y0;
+    xOff = ctx.xOff;
+    yOff = ctx.yOff;
+  }
```

Comment:
The equivalent code in the hardware backends looks like this:

https://github.com/dolphin-emu/dolphin/blob/e92e90d1474d028ba63a3e11b5528e30ebc72c85/Source/Core/VideoCommon/VertexManagerBase.cpp#L574-L592

Apart from some of the excessive minus signs and other confusion that also originally existed in `InitSlope`, there's a divide by zero check that doesn't exist in the software renderer.  Does this actually matter?

The check was added in commit add59b3bea032e331363dcf28361f8e0de65e43d, but https://github.com/dolphin-emu/dolphin/pull/1812#discussion_r22443801 says it shouldn't happen with zfreeze specifically.

Reply:
I think my comment is saying this based on the fact that games that use zfreeze should be sane enough to use non-degenerate triangles for their z reference planes.

No idea what happens on real hardware. These are zero-sized triangles that won't even get rasertized, right? Which is why they don't have a valid slope. It's entirely possible they have already been discarded by this point in the pipeline.